File: 48-rotate-image/rotate-image.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        // Transpose
        for (int i=0; i<matrix.size(); i+=1){
            for (int j=0; j<=i; j+=1){
                swap(matrix[i][j], matrix[j][i]);
            }
        }
        // Reverse each row
        for (int i=0; i<matrix.size(); i+=1){
            reverse(matrix[i].begin(), matrix[i].end());
        }
    }
};

File: 48-rotate-image/rotate-image.go
func rotate(matrix [][]int)  {
    n := len(matrix)
    for i := 0; i < n; i++ {
        for j := 0; j <= i; j++ {
            matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
        }
    }

    for i := 0; i < n; i++ {
        row := matrix[i]
        reverse(row)
        matrix[i] = row
    }
}
func reverse(row []int) {
    i, j := 0, len(row)-1
    for i < j {
        row[i], row[j] = row[j], row[i]
        i++
        j--
    }
}